File: Leet_Code/Convert_Sorted_Array_to_Binary_Search_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def helper(self, nums: List[int], low: int, high: int) -> Optional[TreeNode]:
        if low == high:
            return None
        mid = (low+high)//2
        head = TreeNode(nums[mid])
        head.left = self.helper(nums, low, mid)
        head.right = self.helper(nums, mid+1, high)
        return head

    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        left, right = 0, len(nums)
        mid = (left+right)//2
        head = TreeNode(nums[mid])
        head.left = self.helper(nums, left, mid)
        head.right = self.helper(nums, mid+1, right)
        return head

File: Leet_Code/Convert_Sorted_Array_to_Binary_Search_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode helper(int[] nums, TreeNode node, int left, int right) {
        if (left > right) {
            return null;
        }
        int mid = (int)(left+right)/2;
        TreeNode sub_node = new TreeNode(nums[mid]);
        sub_node.left = helper(nums, sub_node.left, left, mid-1);
        sub_node.right = helper(nums, sub_node.right, mid+1, right);
        return sub_node;
    }
    public TreeNode sortedArrayToBST(int[] nums) {
        int left = 0, right = nums.length-1;
        int mid = (int)(left+right)/2;
        TreeNode root = new TreeNode(nums[mid]);
        root.left = helper(nums, root.left, left, mid-1);
        root.right = helper(nums, root.right, mid+1, right);
        return root;
    }
}
